test/testutils/validators.go

```diff
@@ -55,31 +55,67 @@ func ExpectLeaderSetExist(ctx context.Context, lws *leaderworkerset.LeaderWorker
 		var headlessService corev1.Service
-		if err := k8sClient.Get(ctx, types.NamespacedName{Name: lws.Name, Namespace: lws.Namespace}, &headlessService); err != nil {
+		var headlessServiceList corev1.ServiceList
+		if err := k8sClient.List(ctx, &headlessServiceList, client.InNamespace(lws.Namespace)); err != nil {
 			return false, err
 		}
-		// we expect exactly one service and it must be of the correct type.
-		if headlessService.ObjectMeta.Name != lws.Name {
-			return false, errors.New("service name mismatch")
+		if lws.Spec.NetworkConfig.SubdomainPolicy == leaderworkerset.SubdomainShared {
+			if len(headlessServiceList.Items) != 1 {
+				return false, fmt.Errorf("expected 1 headless service, got %d", len(headlessServiceList.Items))
+			}
+			if err := k8sClient.Get(ctx, types.NamespacedName{Name: lws.Name, Namespace: lws.Namespace}, &headlessService); err != nil {
+				return false, err
+			}
+			return validateService(headlessService, lws, lws.Name, map[string]string{leaderworkerset.SetNameLabelKey: lws.Name})
 		}
-		if headlessService.ObjectMeta.Namespace != lws.Namespace {
-			return false, errors.New("service namespace mismatch")
+
+		if len(headlessServiceList.Items) != (int(*lws.Spec.Replicas) + 1) {
+			return false, fmt.Errorf("expected %d headless services, got %d", (int(*lws.Spec.Replicas) + 1), len(headlessServiceList.Items))
 		}
-		if headlessService.Spec.ClusterIP != "None" {
-			return false, errors.New("service type mismatch")
+
+		if err := k8sClient.Get(ctx, types.NamespacedName{Name: lws.Name, Namespace: lws.Namespace}, &headlessService); err != nil {
+			return false, err
 		}
-		if !headlessService.Spec.PublishNotReadyAddresses {
-			return false, errors.New("service publish not ready should be true")
+
+		if valid, err := validateService(headlessService, lws, lws.Name, map[string]string{leaderworkerset.PodRoleLabelKey: "leader"}); err != nil {
+			return valid, err
 		}
-		selector := headlessService.Spec.Selector
-		value, exists := selector["leaderworkerset.sigs.k8s.io/name"]
-		if !exists || value != lws.Name {
-			return false, errors.New("selector name incorrect")
+
+		for i := 0; i < int(*lws.Spec.Replicas); i++ {
+			if err := k8sClient.Get(ctx, types.NamespacedName{Name: fmt.Sprintf("%s-%s", lws.Name, strconv.Itoa(i)), Namespace: lws.Namespace}, &headlessService); err != nil {
+				return false, err
+			}
+			if _, err := validateService(headlessService, lws, fmt.Sprintf("%s-%s", lws.Name, strconv.Itoa(i)), map[string]string{leaderworkerset.PodRoleLabelKey: "worker", leaderworkerset.GroupIndexLabelKey: strconv.Itoa(i)}); err != nil {
+				return false, err
+			}
 		}
 		return true, nil
 	}, Timeout, Interval).Should(gomega.Equal(true))
 }
 
+func validateService(headlessService corev1.Service, lws *leaderworkerset.LeaderWorkerSet, serviceName string, wantSelector map[string]string) (bool, error) {
+	if headlessService.ObjectMeta.Name != serviceName {
+		return false, errors.New("service name mismatch")
+	}
```

Comment:
consider use this approach to print out the diff for easier debugging? https://github.com/kubernetes-sigs/lws/blob/93bfa66aa718c05ba4a4dcd2082df9e05b480b2e/test/testutils/validators.go#L138C3-L143C19